### ISSUE TITLE
How hard negative contrastive loss work?

### ISSUE THREAD

**Issue by voorhs** [amazon-research/dse#8]
I am trying to understand this [code snippet](https://github.com/amazon-science/dse/blob/f645d74efd8b6a8b12593f9294f1d482e42946ae/pretrain/utils/contrastive_utils.py#L46C9-L51C26):
```
negimp = neg.log().exp()
Ng = (negimp*neg).sum(dim = -1) / negimp.mean(dim = -1)
loss_pos = (-posmask * torch.log(pos / (Ng+pos))).sum() / posmask.sum()
```

As it seems to me, this code implements not the same function as was introduced in paper. It implements this (using [paper](https://assets.amazon.science/fe/f3/aedb0d0746ad8ee21c5d9693bfd7/learning-dialogue-representations-from-consecutive-utterances.pdf)'s notation):
```math
\ell^{i,i^+}=-\log\frac{\exp(\text{sim}(e_i,e_{i^+})/\tau)}{\displaystyle\exp(\text{sim}(e_i,e_{i^+})/\tau)+\sum_{j=1,j\neq i,j\neq i^+}^{2M}\alpha_{ij}\cdot\exp(\text{sim}(e_i,e_j)/\tau)},\quad \alpha_{ij}=\frac{\exp(\text{sim}(e_i,e_j)/\tau)}{\displaystyle\frac{1}{2M-2}\sum_{k=1,k\neq i,k\neq i^+}^{2M}\exp(\text{sim}(e_i,e_k)/\tau)}
```
where $e_1,\ldots,e_M$ and $e_{M+1},\ldots,e_{2M}$ are the context and response encodings respectively ($i^+:=i+M$). But according to paper, it doesn't seem to match.

Am I misinterpreting paper or/and code?
